Developed GUI 1.0

nh916 opened the pull request:
# Description

The GUI developed here will take the place of the config.yaml file and the terminal previously used by the user to
upload Excel file to CRIPT. The GUI will have a place for the user to enter their needed credentials to connect and
upload to the CRIPT platform, show a loading screen and progress of what is being uploaded, and an error screen
displaying the Excel errors for the user to fix.

# Changes

Excel uploader original code is located in src/python/ with minimal changes to all files except for what used to be
called main.py, that is now refactored into a class called excel_uploader_main.py. All the python files including
gui_main.py can use some refactoring to improve them for the future

The package being used to make the GUI is called [eel](https://github.com/python-eel/Eel). I am using html, css,
bootstrap css, and plain JS to run it in Chrome app mode. The python file that controls the GUI is gui_main.py and all
UI code is located in src/web/

For the graphics I am downloading them from [undraw.co](https://undraw.co/), and placing them in src/web/assets/
The condensed CRIPT logo used for the executable icon is placed in src/assets/

Furthermore, README.md has also been edited to include attribution badges to show what technologies we are using

## Excel template changes
The Excel template has been changed to include formulas and validation for **Row 1**(category), **Row 2**(Column Name), **Row 3**(units) _and now a new row_ **Row 4**(instructions). Thus, the Excel Uploader considers row 1-4 to be headers and starts reading user data from row 5. The new example Excel file is: [CRIPT_template_Simplified EXAMPLE.xlsx](https://github.com/C-Accel-CRIPT/cript-excel-uploader/files/9865586/CRIPT_template_Simplified.EXAMPLE.xlsx) based on the old example Excel file: [example_template_v0-4-2.xlsx](https://github.com/C-Accel-CRIPT/cript-excel-uploader/releases/download/v0.4.2/example_template_v0-4-2.xlsx)

# Tests

I have not written any unit or integration tests for this program yet. I have tested it manually only on a Windows 11
computer so far using a
single [Excel example template](https://github.com/C-Accel-CRIPT/cript-excel-uploader/releases/download/v0.4.2/example_template_v0-4-2.xlsx)

# Known Bugs

* The major bug observed so far is GUI can be closed but the program can still be running in memory and has to be
  stopped from task manager manually
* If the user does not have chrome installed on their computer the program cannot run. Chrome must be installed for the
  executable to run. This is a requirement by the [eel](https://github.com/python-eel/Eel) package
* There may be some exceptions that are not handled that might be able to break the program. More testing is needed to
  know for sure
* Very heavily integrated with Globus and not integrated or tested with any other storage clients

# How to compile to an executable

## Windows

The way that I have used to compile in the past has been with the following steps

1. Create and activate virtualenv
    1. `python -m venv .\venv`
    2. ` .\venv\Scripts\activate`
2. install dependencies from `requirements.txt`
    1. `pip install -r requirements.txt`
3. compile code to executable
    1. `cd .\src\`
    2. ` python -m eel .\gui_main.py .\web\ --onefile --noconsole --name excel_uploader_gui_windows --icon=.\assets\logo_condensed.ico`
        1. Note: If you are compiling multiple times and running into issues, please be sure to clean out anything that
           might have been cached that could cause issues. You can try clearing Chrome cache, deleting venv and
           reinstalling dependencies, remove any problematic files or folders
           from `C:\Users\myusername\AppData\Local\Temp` and using [CCleaner](https://www.ccleaner.com/)
    3. This should create a build/ directory, dist/ directory and an excel_uploader_gui_windows.spec file
    4. The GUI executable is located in dist/excel_uploader_gui_windows.exe

# CD

There are some GitHub actions written for compiling python code to an executable for windows, Linux (Ubuntu), and Mac.
The only one that has been worked on the most has been the windows action, and it is not fully done. The other versions
have been written, but not tested and will likely need some work to get them fully functional.

The actions can be found in:

* `.github/workflows/pyinstaller_gui_mac.yaml`
* `.github/workflows/pyinstaller_gui_ubuntu.yaml`
* `.github/workflows/pyinstaller_gui_windows.yaml`

nh916 committed:
Merge remote-tracking branch 'origin/master'

nh916 committed:
adding GUI to repo

nh916 committed:
removing ascii art as it is no longer needed

nh916 committed:
removing footer from base.html as it is not currently needed

nh916 committed:
removed {% block footer %} from base.html

nh916 committed:
base.html: put header inside the header tag

nh916 committed:
put all screens into a single html and made routes.js to work with it

gui_main.py: runs base.html instead of start_screen.html

routes.js: working similarly but the inside code has changed to show and hide screens

base.html: contains all the screens

.gitignore: dev.js for quick fill is kept private

nh916 committed:
can show errors to screen

nh916 committed:
added version to title for users to know which executable they are using

nh916 committed:
hooked up try again button

nh916 committed:
giving out error_screen.js with traceback, but the errors can be way too much

nh916 committed:
added comments to error_screen.js

nh916 committed:
gui working similar to the original excel_uploader_main.py

nh916 committed:
using a giant upload_driver that has the exact code as original excel uploader main.py

base.html: added span element to show what is being uploaded

base.html: removed error for data privacy options radio buttons because radio buttons can't really have errors

loading_screen.js: have the code ready to show what is being uploaded

loading_screen.js: using python to navigate to success screen and not when js hits 100% uploaded

excel_uploader_main.py: upload_driver() is exactly the code from original excel uploader main.py

excel_uploader_main.py: upload_to_cript function not currently working or being called

upload.py: putting it back as it originally was for now

nh916 committed:
loading screen progress bar working

base.html: changed progress label from span to em because that reads better

excel_uploader_main.py: implemented idea of progress bar with uploads and upload.py and

upload.py: using current_progress, total_progress,

 gui_main.py: updated doc string for update_progress_bar, so it can be easily understood

nh916 committed:
excel_uploader_main.py: removed get_parsed_sheet(), create_nodes(), upload_to_cript()

because currently using a giant upload_driver() instead

nh916 committed:
built out globus authentication screen in base.html

base.html: built globus authentication screen

routes.js: added goToGlobusAuthScreen() and made code more dry with the addition of hideAllScreens()

globus_screen.js: added js file with function to set the globus url

added secure_login.png for globus auth screen

gui_main.py: after send_to_upload() will call globus_auth() method

excel_uploader_main.py: put get collections url in its own method that can be called easily

nh916 committed:
when user closes the app window, the program exits, and added some doc strings to upload.py

gui_main.py: implemented user_closed_app()

all_screens.js: when user closes the app, I call python function to exit the program

base.html: added all_screens.js to head
base.html: removed jinja code because not using jinja anymore

excel_uploader_main.py: retuning at the end of upload part of upload_driver()

upload.py: added doc strings
upload.py: formatted

nh916 committed:
updated requirements.txt and requirements_dev.txt

base.html: commented out dev.js for quickFill

nh916 committed:
updated requirements for users and dev

nh916 committed:
globus screen and js and python functions built out and password visibility toggle works for globus screen and start screen

start_screen.js: added api token visibility toggle

gui_main.py: added placeholder methods for globus

gui_main.py: added doc strings to globus methods

globus_screen.js: added globus token visibility toggle and can submit from frontend to python eel

nh916 committed:
base.html some styling on start screen

nh916 committed:
works correctly by taking users to error screen when they get an error

nh916 committed:
catching cript.exceptions.UnsavedNodeError in upload.py and showing it to the user, but not really handling it

nh916 committed:
progressbar resetting from error_screen.js

nh916 committed:
progress bar working correctly as it used to

nh916 committed:
progress bar working correctly and upload.py upload() setting the variable instead of returning anything

nh916 committed:
upload.py: if statement for if using files from web instead of local then dont ask for globus authentication

nh916 committed:
going to error screen or globus screen correctly

nh916 committed:
removed Try again button in error screen and error list is in excel_uploader_main.py

base.html: commented out try again button for now because the errors persisting in the program can mess it up, and that can be fixed later

styles.css: expanded error-window since there is no Try again button below it

excel_uploader_main.py: maintaining the error_list

nh916 committed:
gui works with success_screen

nh916 committed:
gui can upload successfully

nh916 committed:
gui start screen input labels are all uniform width

nh916 committed:
updated requirements_dev.txt

nh916 committed:
gui_main.py fixed the path issue so it will work regardless of where in the terminal you are

nh916 committed:
removed print statement

nh916 committed:
optimized imports and formatted gui_main.py

nh916 committed:
pyinstaller_compile.bash script added for compiling code from python with pyinstaller on windows to executable

nh916 committed:
Merge branch 'master' of https://github.com/C-Accel-CRIPT/cript-excel-uploader

nh916 committed:
removed all other html templates besides base.html because other templates will no longer be used

nh916 committed:
updated title to be alpha version

nh916 committed:
added instructions to error screen and because of that styles.css and base.html also changed

nh916 committed:
progress bar is animated

nh916 committed:
data will be kept non-public

base.html: removed private and public option and added "your data will be uploaded as non-public"

start_screen.js: removed js looking for private and public options

gui_main.py: removed setting data privacy and just hard coding it to private

nh916 committed:
changed success screen button to CRIPT purple

nh916 committed:
made "Authentication Failure" more bold

nh916 committed:
added python exit handler so when python is closes the gui closes as well

nh916 committed:
changed extension for pyinstaller_compile.sh

nh916 committed:
removed !bash from pyinstaller_compile.sh

nh916 committed:
changed the path for html_path, minor change

nh916 committed:
removed finished TODO about renaming alternative_main.py

nh916 committed:
visibility toggle button added changing icons to give feedback of which state it is in

nh916 committed:
added JS comments for documentation

nh916 committed:
added bash file to compile on windows and ubuntu

nh916 committed:
added bash header to pyinstaller_compile_ubuntu.sh

nh916 committed:
added github action files for compiling gui in mac, windows, and linux

nh916 committed:
added name flag to pyinstaller_compile_windows.sh

nh916 committed:
fixed html_path variable

nh916 committed:
added visibility icon switching for globus auth screen

nh916 committed:
updated create.py and parse.py to what is in develop branch

nh916 committed:
changed GUI port to 8001, so it does not clash with anything else running

nh916 committed:
worked on globus authentication. initial version is working and can authenticate with globus

gui_main.py: built out the globus_auth() method logic

gui_main.py: updated doc strings

excel_uploader_main.py: updated doc strings

excel_uploader_main.py: changed upload.upload part of upload_driver to reflect updates in GitHub dev branch

excel_uploader_main.py: built out set_globus_auth_link() and is_globus_auth_token_valid()

base.html: removed dev.js script import

base.html: start screen says "Your data will be uploaded as unpublished"

base.html: globus screen added accessibility aria lables

base.html: added validation feedback to globus screen

globus_screen.js: added validation feedback function to globus_screen.js

nh916 committed:
redesigned UI for Globus Auth Screen

nh916 committed:
added todo to gui_main.py

nh916 committed:
takes user to globus correctly if they only have a local file to upload

nh916 committed:
removed jinja as a dependency in requirements_dev.txt

nh916 committed:
updated requirements.txt and requirements_dev.txt

nh916 committed:
GUI compiling successfully, and fixed the html path variable

nh916 committed:
removed tqdm from code and requirements_dev.txt

nh916 committed:
removed print statements used for debugging

nh916 committed:
program compiles successfully

nh916 committed:
downgraded the requirements into what it was

nh916 committed:
excel_uploader_main.py: updated is_uploading_local_files to use .startswith() instead of just checking if a http is in the whole string

nh916 committed:
gui_main.py: initializing eel inside the constructor

nh916 committed:
loading_screen.js: stopping progress bar from going over 100%

rjmello committed:
Fix data sheet parameters

nh916 committed:
excel_uploader_main.py: fixing get_total_for_progress_bar( ) because at the end I added 3 to debug but I forgot to remove it

nh916 committed:
added attribution shields to README.md

nh916 committed:
fixed material-mkdocs badge

nh916 committed:
changed styling for attributions

nh916 committed:
updated docs link heading in README.md

nh916 committed:
changing start screen back to non-public

nh916 committed:
removed the commented out tkinter dialog box cript icon

nh916 committed:
added tkinter.destroy() if either the program is closed or python exits

nh916 committed:
saving if the dialog box is already open or not, so user cannot open a 100 instances of a dialog box

nh916 committed:
removed some solved TODO comments

nh916 committed:
formatted src/python/ with black

nh916 committed:
removed the terminal command files to compile the GUI to executable

nh916 committed:
putting requirements_dev.txt into requirements.txt and removing requirements_dev

This is because users no longer need to pull the code and run a requirements.txt file to run python from their terminal. Instead the only people installing requirements will be developers or advanced users

nh916 committed:
resolving merge conflicts between master and develop-GUI

the main conflicts were that I had moved create.py and parse.py into src/python/ and git did not recognize that

requirements.txt was also in conflict because in master it was 0.4.7, which was outdated and we needed to use 0.5.0

nh916 committed:
removed (alpha) from the top of the GUI since it will be distributed to users and is now versioned as 1.0

rjmello commented on .github/workflows/pyinstaller_gui_mac.yaml:
```diff
@@ -0,0 +1,25 @@
+name: Compile GUI for Mac
```
Let's hold off on deploying workflows to master that haven't been tested. We can chat about the various ways to accomplish this on the side.

nh916 replied:
I have tested this on the other [repo](https://github.com/nh916/tkinter-and-eel) that I was originally working on
I have the action to run `"on: workflow_dispatch"`, which means that this action only runs when I manually go into GitHub and and click run

rjmello commented on requirements.txt:
```diff
@@ -1,5 +1,7 @@
+black==22.3.0
+pre-commit==2.17.0
+eel==0.14.0
+pyinstaller==5.4.1
```
Separating which dependencies are required for the code to function from others is quite useful. I don't see a reason to deliberately move away from this, so let's recreate the `requirements_dev.txt` file.

rjmello commented on src/gui_main.py:
```diff
@@ -0,0 +1,311 @@
+import requests
+
+# my imports
+from python.excel_uploader_main import ExcelUploader
```
Minor detail: there's no need to add comments describing the types of imports, since this grouping is described in [PEP 8](https://peps.python.org/pep-0008/#imports).

rjmello commented on src/gui_main.py:
```diff
@@ -0,0 +1,311 @@
+        self.api_key = None
+        self.project_name = None
+        self.collection_name = None
+        self.data_is_public = False
```
We will not be allowing users to upload data as public from this tool, so this can be removed entirely.

rjmello reviewed (changes requested):
This is very much still in testing phase, so we should bump the version to 0.5.0, not 1.0.0.

nh916 committed:
resolving requested changes to merge into master branch

base.html: changed from v1.0 to v0.5

gui_main.py: removed comments about imports

requirements.txt: removed black, eel, pyinstaller, and pre-commit and placed them inside of requirements_dev.txt

excel_uploader_main.py: removed self.data_is_public from __init__()

nh916 committed:
resolving requested changes to merge into master branch

removing unused variables from gui_main.py: host, api_key, project_name, collection_name, data_is_public

nh916 committed:
adding development branch changes into this branch

nh916 committed:
fixed version typo

nh916 committed:
cleaned up assets and putting images in /assets/images/ folder

nh916 committed:
jazzed up loading screen: added description of things that the program does

nh916 committed:
updated README.md badges to include color, logo, and spacing them out

nh916 committed:
removing space between badges

nh916 committed:
reworded badges from "packages" to "dependencies" and fixed typos

nh916 committed:
Merge remote-tracking branch 'origin/develop-GUI' into develop-GUI

# Conflicts:
#	README.md

nh916 committed:
reworded badges from "packages" to "dependencies" and fixed typos

nh916 committed:
spaced out badges for better readability

nh916 committed:
removed the extra space. looks better without it

nh916 committed:
removed duplicate description badge for loading screen

nh916 committed:
"Try Again" button works correctly from the Error screen

base.html: enabled the "Try Again" button and removed the instructional text about telling the user to "close the program and relaunch it"

error_screen.js: created the functions to reset all screens and then call python to clear the error_list and set current_progress to 0

gui_main.py: created a reset_for_uploading() method that empties out the error_list and then calls excel_uploader_main.py to clear out the current_progress for progress bar

nh916 committed:
when clicking the "Try Again" button from error_screen, it will clear out any feedback and validation from start_screen so when navigating back there are not any left over feedback

nh916 committed:
updating GUI to newest SDK version, but getting errors and more work is needed

add_sample_preparation_to_process() seems to be broken. Getting an error of:

cript.data_model.exceptions.UnsavedNodeError: Experiment nodes must be saved before they can be added to other nodes.

nh916 committed:
updating GUI to newest SDK version, but getting an error when adding sample_preparation:

cript.data_model.exceptions.UnsavedNodeError: Process nodes must be saved before they can be added to other nodes.

nh916 committed:
skipping Reference Node duplicate uploads

upload.py: if trying to save Reference node that already exists then the program just skips it and continues uploading everything else

excel_uploader_main.py: just removed a comment that is not needed anymore

nh916 committed:
added 5 ms of pause between each upload to keep our server safe from DDOS

brili commented on src/python/upload.py:
```diff
@@ -0,0 +1,102 @@
+
+            # update progress bar regardless of what happens
+            update_progress_bar(obj_type, excel_uploader_object, gui_object)
+
```
Time.sleep needs to go here, so it is inside the for loop

nh916 replied:
I added a 5 ms sleep between the uploads in upload.py in the last commit

nh916 committed:
remove the pause from between upload calls within excel_uploader_main.py and placed it inside of upload.py between each upload to cript to prevent any DDOS

nh916 committed:
changing headers to rows 1-4 because of new Excel file has 4th row as instructions/description and the value starts at row 5
